[DOCS] Adds inference processor documentation (#50204)
elastic/elasticsearch: docs/reference/ingest/processors/inference.asciidoc

[role="xpack"]
[testenv="basic"]
[[inference-processor]]
=== {infer-cap} Processor

Uses a pre-trained {dfanalytics} model to infer against the data that is being 
ingested in the pipeline.

[[inference-options]]
.{infer-cap} Options
[options="header"]
|======
| Name               | Required  | Default                        | Description

> | `model_id   `      | yes       | -                              | (String) The ID of the model to load and infer against.

Reviewer: This formatting isn't working in the preview.  Not sure if this will fix the problem?:
```suggestion
| `model_id`      | yes       | -                              | (String) The ID of the model to load and infer against.
```

> | `target_field`     | no        | `ml.inference.<processor_tag>` | (String) The field name to which to write the results objects.

Reviewer: I like the way this is worded in the enrich processor.  Can we do something similar? For example:
```suggestion
| `target_field`     | no        | `ml.inference.<processor_tag>` | (String) Field added to incoming documents to contain results objects.
```

| `field_mappings`   | yes       | -                              | (Object) Maps the document field names to the known field names of the model.

> | `inference_config` | yes       | -                              | (Object) Contains the inference type and its options. There are two types: `regression` and `classification`.

Reviewer: Should this include links to the later sections?
```suggestion
| `inference_config` | yes       | -                              | (Object) Contains the inference type and its options. There are two types: <<inference-processor-regression-opt,`regression`>> and <<inference-processor-classification-opt,`classification`>>.
```

include::common-options.asciidoc[]
|======

[source,js]
--------------------------------------------------
{
  	"inference": {
      	"model_id": "flight_delay_regression-1571767128603",
      	"target_field": "FlightDelayMin_prediction_infer",
      	"field_mappings": {},
      	"inference_config": {"regression": {}},
      	"model_info_field": "ml"
    }
}
--------------------------------------------------
// NOTCONSOLE

[discrete]
[[inference-processor-regression-opt]]
==== {regression-cap} configuration options

`results_field`::
(Optional, string)
Specifies the field to which the inference prediction is written. Defaults to 
`predicted_value`.

[discrete]
[[inference-processor-classification-opt]]
==== {classification-cap} configuration options

`results_field`::
(Optional, string)

> Specifies the field to which the inference prediction is written. Defaults to 

Reviewer: Related to earlier comment, if applicable, the wording in the enrich processor is nice:
```suggestion
The field that is added to incoming documents to contain the inference prediction. Defaults to 
```

[role="xpack"]
[testenv="basic"]
[[inference-processor]]
=== {infer-cap} Processor

Uses a pre-trained {dfanalytics} model to infer against the data that is being 
ingested in the pipeline.

[[inference-options]]
.{infer-cap} Options
[options="header"]
|======
| Name               | Required  | Default                        | Description
| `model_id   `      | yes       | -                              | (String) The ID of the model to load and infer against.

> | `target_field`     | yes       | `ml.inference.<processor_tag>` | (String) The field name to which to write the results objects.

Reviewer 2: `Required: no` since we have a sane default

> | `field_mappings`   | no        | -                              | (Object) Maps the document field names to the known field names of the model.

Reviewer 2: I think this is required...if testing showed otherwise, I think that should be fixed.

Author: The two required field mistakes are copy/paste errors. Thanks!

> | `inference_config` | yes       | -                              | (Object) Contains the inference type and its options. There are two types: `regression` and `classification`.

Reviewer 2: It would probably be helpful to have full docs referencing all the options for `regression` and `classification` configs. I am not sure where they would fit in this doc format (probably outside the processor field table though).

```
regression:
- results_field: the field to which the inference prediction is written. Defaults to `predicted_value`
```
```
classification:
- results_field: the field to which the inference prediction is written. Defaults to predicted_value
- top_num_classes: The number of top class predictions to return. Defaults to 0.
- top_classes_results_field: The field to which the top classes are written. Defaults to top_classes
```
